pkg/projects/resource_project_user.go

```diff
@@ -0,0 +1,214 @@
+	Equals(other Comparable) bool
+}
+
+func (user User) Hash() string {
+	return user.Name
+}
+
+func (a User) Equals(b User) bool {
+	return a.Hash() == b.Hash()
+}
+
+func contains(as []User, b User) bool {
+	log.Printf("[DEBUG] contains")
+	log.Printf("[TRACE] as: %+v\n", as)
+	log.Printf("[TRACE] b: %+v\n", b)
+
+	for _, a := range as {
+		log.Printf("[TRACE] a: %+v\n", a)
+		log.Printf("[TRACE] a.Equals(b): %+v\n", a.Equals(b))
+		if a.Equals(b) {
+			return true
+		}
+	}
+	return false
+}
+
+// Not the most efficient way to determine the slices intersection but this suffices for the small-ish number of items
+var usersIntersection = func(as []User, bs []User) []User {
+	log.Printf("[DEBUG] usersIntersection")
+	log.Printf("[TRACE] as: %+v\n", as)
+	log.Printf("[TRACE] bs: %+v\n", bs)
+
+	intersection := make([]User, 0)
+
+	for _, a := range as {
+		if contains(bs, a) {
+			intersection = append(intersection, a)
+		}
+	}
+
+	return intersection
+}
+
+// Not the most efficient way to determine the slices differences but this suffices for the small-ish number of items
+var usersDifference = func(as []User, bs []User) []User {
+	difference := make([]User, 0)
+
+	for _, a := range as {
+		if !contains(bs, a) {
+			difference = append(difference, a)
+		}
+	}
+
+	return difference
+}
+
+type Users struct {
+	Members []User
+}
+
+const projectUsersUrl = "/access/api/v1/projects/%s/users/"
+
+var unpackUsers = func(data *schema.ResourceData) (string, Users, error) {
+	d := &ResourceData{data}
+	projectKey := d.getString("key", false)
+
+	users := Users{
+		Members: make([]User, 0),
+	}
+
+	if v, ok := d.GetOkExists("user"); ok {
+		projectUsers := v.(*schema.Set).List()
+		if len(projectUsers) == 0 {
+			return projectKey, users, nil
+		}
+
+		for _, projectUser := range projectUsers {
+			id := projectUser.(map[string]interface{})
+
+			user := User{
+				Name:  id["name"].(string),
+				Roles: castToStringArr(id["roles"].(*schema.Set).List()),
+			}
+			users.Members = append(users.Members, user)
+		}
+	}
+
+	return projectKey, users, nil
+}
+
+var packUsers = func(d *schema.ResourceData, key string, users *[]User) []error {
+	log.Printf("[DEBUG] packUsers")
+
+	setValue := mkLens(d)
+
+	var projectUsers []interface{}
+
+	for _, user := range *users {
+		log.Printf("[TRACE] %+v\n", user)
+		projectUser := map[string]interface{}{
+			"name":  user.Name,
+			"roles": user.Roles,
+		}
+
+		projectUsers = append(projectUsers, projectUser)
+	}
+
+	log.Printf("[TRACE] %s\n", key)
+	log.Printf("[TRACE] %+v\n", projectUsers)
+
+	errors := setValue(key, projectUsers)
+
+	return errors
+}
+
+var getProjectsUsersUrl = func(projectKey string, id string) string {
+	return fmt.Sprintf(projectUsersUrl, projectKey) + id
+}
+
+var readUsers = func(projectKey string, m interface{}) ([]User, error) {
+	log.Println("[DEBUG] readUsers")
+
+	users := Users{}
+
+	_, err := m.(*resty.Client).R().SetResult(&users).Get(fmt.Sprintf(projectUsersUrl, projectKey))
```

Comment:
Resty has built in templating for the user `SetParameter("thing","value")` - it's more idiomatic